src/tracerunner.rs

```diff
@@ -0,0 +1,159 @@
+            }
+        }
+
+        Runner {
+            interp: interp,
+            trace: &trace.trace,
+            stack: Vec::new(),
+            locals: locals,
+        }
+    }
+
+    pub fn run(&mut self) -> InstrPtr {
+        use TraceInstruction::*;
+
+        let mut pc = 0;
+        loop {
+            let instr = &self.trace[pc];
+            pc = (pc + 1) % self.trace.len();
+
+            info!("TEXEC: {:?}", instr);
+
+            match *instr {
+                Add        => self.add(),
+                Cmp(how)   => self.cmp(how),
+                Load(idx)  => self.load(idx),
+                Store(idx) => self.store(idx),
+                ArrayGet   => self.array_get(),
+                Const(val) => self.stack.push_from(val),
+                Clone      => {}
+
+                Guard(ref guard) => {
+                    if let Err(recovery) = self.check_guard(guard) {
+                        return recovery;
+                    }
+                }
+
+                // these opcodes are not needed for example
+                Array(_) | Push | Print | Len => unimplemented!(),
+            }
+        }
+    }
+
+    fn check_guard(&mut self, guard: &Guard) -> Result<(), InstrPtr> {
+        let check = self.stack.pop_into::<bool>() == guard.condition;
+        check.ok_or_else(||{
+                self.recover(guard);
+                InstrPtr::new(guard.frame.func.clone(), guard.pc)
+            })
+    }
+
+    /// Recovery (aka Blackholing)
+    ///
+    /// Execution has reached a point, where the trace isn't valid anymore.
+    /// The goal is to return to the interpreter, but the state has to be
+    /// recovered first.
+    ///
+    /// The following states have to be recovered:
+    ///     * stack-frames (call-frames)
+    ///       The failed guard might have failed within an inlined function call.
+    ///       Thus, we have to reconstruct all missing callframes, before the
+    ///       the interpreter can gain back control.
+    ///       Second, we also have to consider the frame where the loop resides
+    ///       in, since state might has also has changed there.
+    ///
+    ///     * value stack
+    ///       Also the operand stack has to be recovered.
+    ///       Foremost, the condition, which caused the guard to fail, has to be
+    ///       restored.
+    ///       TODO: Are there other values which might have to be recovered?
+    fn recover(&mut self, guard: &Guard) {
+        // remove the last callframe of the Interpreter
+        // it gets replaced with our updated version
+        self.interp.frames.pop().unwrap();
+
+        // recover callframes
+        // since callframes depend on each other, we start with the one which
+        // was created first
+        let frames = guard.frame.walk().collect::<Vec<_>>();
+
+        // why .rev()? We rebuild the stack of frames, so we have to start with
+        // the bottom most.
```

Comment:
Maybe it's better to talk about a concrete example.

```
s = Stack()
s.push(1)
s.push(2)
s.push(3)
```

In my view, `1` is on the bottom of the stack, so it is the bottom-most value. The top of the stack is the last value added, in this case `3`. What I want to say in the comment is that we have to start with `1` to rebuild the stack.